### ISSUE TITLE
[BUG] Blink drawer when open the app

### ISSUE THREAD

**Issue by djorkaeffalexandre** [react-navigation/react-navigation#7561]
I tried to upgrade my `react-navigation-drawer` to `2.2.1` version but when I open the app it blinks a sidebar..

### GIF
![](https://media.giphy.com/media/hS3ISjpGvcBlC3Ze24/giphy.gif)

You can see the code here:
https://github.com/RocketChat/Rocket.Chat.ReactNative/blob/develop/app/index.js

**Comment by iamyellow** [comment]
just in case it helps somebody... what i've done is using the `progress` prop the drawer component gets. actually, it's a `reanimated` node which goes from 0-1, so we can use it like this:
```
const opacity = useMemo(() => {
  const { block, cond, eq } = Animated
  return block([cond(eq(props.progress, 0), 0, 1)])
}, [])
```
so, the root component for my drawer goes like this:
````
<Animated.View style={[styles.root, { opacity }]}>
````
where `styles.root` is whatever style you want, like `{ flex: 1, backgroundColor: 'white' }`

keep in mind, this is just a hack / workaround until the issue is being solved.

**Comment by okwast** [comment]
We actually had the same issue in React Navigation 5.

We use these versions:
```
"@react-navigation/bottom-tabs": "5.3.3",
"@react-navigation/drawer": "5.6.3",
"@react-navigation/native": "5.2.3",
"@react-navigation/stack": "5.2.18",
"react-native": "0.61.5",
```

We fixed it for now by using a hook to check for the initial render.
```
export const useInitialRender = (): boolean => {
  const ref = useRef(false);

  if (!ref.current) {
    ref.current = true;
    return true;
  }
  return false;
};
```
In the navigator we then set the width of the drawer to `0` for the initial render.
```
<Drawer.Navigator
      overlayColor={...}
      drawerContent={drawerContent}
      // Width is set to 0 for initial render to prevent a flicker
      drawerStyle={{ width: isInitialRender ? 0 : drawerWidth }}
      initialRouteName="Tabs"
    >
      <Drawer.Screen
        name="Tabs"
        component={TabNavigator}
      />
    </Drawer.Navigator>
```
Note: Setting the width to `0` seemed to be even nicer than rendering null for the content in the initial render.

If we are the only ones having this issue with ReactNavigation 5 we probably messed something up though 🤔

**Comment by zachschultz** [comment]
@okwast I've got the same issue with `react-navigation/drawer@5.8.0` and `react-navigation/native@5.4.3`.

**Comment by raajnadar** [comment]
Can anyone add a repro to debug??

**Comment by Ubanna** [comment]
I have the same issue too with navigation v5 ---  "@react-navigation/drawer": "^5.8.0"

**Comment by raajnadar** [comment]
Closing since no minimal runnable repro was provided! If anyone can provide a repro create a new issue with more info.

I tested 4 project using react navigation v5 and could not reproduce it. Try upgrading to version 5 maybe your issue will get fixed.

**Comment by sauravsid** [comment]
i have this issue in RN 5

**Comment by Bryan-b** [comment]
solved
just declare a variable then set it to false.
the use the variable conditionally in your  'drawerStyle'

`const showDrawer = false;`
`drawerStyle={{width: !showDrawer ? null : 280}}`

**Comment by xXFracXx** [comment]
Still facing this issue.

**Comment by silentashish** [comment]
> Solved!
> i'm using react navigation version 5 and here is how i solved mine...
> Just declare a variable then set it to false.
> then use the variable conditionally in Navigation's configuration
> 
> `const showDrawer = false;`
> `drawerStyle={{width: !showDrawer ? null : 280}}`
> 
> !Do not replace `null` with `0` else your drawer width remains `0`

This work in v5. Save my day.

**Comment by heroshinn** [comment]
> We actually had the same issue in React Navigation 5.
> 
> We use these versions:
> 
> ```
> "@react-navigation/bottom-tabs": "5.3.3",
> "@react-navigation/drawer": "5.6.3",
> "@react-navigation/native": "5.2.3",
> "@react-navigation/stack": "5.2.18",
> "react-native": "0.61.5",
> ```
> 
> We fixed it for now by using a hook to check for the initial render.
> 
> ```
> export const useInitialRender = (): boolean => {
>   const [isInitialRender, setIsInitialRender] = useState(false);
> 
>   if (!isInitialRender) {
>     setTimeout(() => setIsInitialRender(true), 1);
>     return true;
>   }
>   return false;
> };
> ```
> 
> In the navigator we then set the width of the drawer to `0` for the initial render.
> 
> ```
> <Drawer.Navigator
>   overlayColor={...}
>   drawerContent={drawerContent}
>   // Width is set to 0 for initial render to prevent a flicker
>   drawerStyle={{ width: isInitialRender ? 0 : drawerWidth }}
>   initialRouteName="Tabs"
> >
>   <Drawer.Screen
>     name="Tabs"
>     component={TabNavigator}
>   />
> </Drawer.Navigator>
> ```
> 
> Note: Setting the width to `0` seemed to be even nicer than rendering null for the content in the initial render.
> 
> If we are the only ones having this issue with ReactNavigation 5 we probably messed something up though 🤔

solved my problem :)

**Comment by jatinderOfc** [comment]
I was able to fix it by doing this:

```javascript
export default function DrawerNavigator() {
  const [initRender, setInitRender] = useState(true);

  useEffect(() => {
    setInitRender(false);
  }, [initRender]);

  return (
    <Drawer.Navigator
      drawerContent={(props) => <AppDrawer {...props} />}
      drawerStyle={{ width: initRender ? null : "90%" }}
    >
        /*screens*/
    </Drawer.Navigator>
```

**Comment by gasbriones** [comment]
@jatinderOfc you are awesome! this works for me! thanks so much!

**Comment by velidan** [comment]
None of the answers above helped me. I'm working with the React Native Web + Next.js
Here is how I solved this issue:

I used the initial render hook @okwast 

```
const useInitialRender = (): boolean => {
  const [isInitialRender, setIsInitialRender] = React.useState(false);

  if (!isInitialRender) {
    setTimeout(() => setIsInitialRender(true), 1);
    return true;
  }
  return false;
};

```

Unfortunately, **width: 0 | null** didn't help so I used this dirty hack. Hope it helps someone.
```
function MyApp({ Component, pageProps }: any) {
  const isInitialRender = useInitialRender();
  
  return (
        <NavigationContainer>
          <Drawer.Navigator 
            openByDefault={false}
            drawerType={ isInitialRender ? 'permanent' : 'front' } // without it the drawer was blinking even with the _display: 'none'_
            drawerStyle={{ 
            width: 240,
            display: isInitialRender ? 'none' : 'flex' }}
            drawerContent={(props: any) => <DrawerContent {...props }/>}
             initialRouteName="Main">
            <Drawer.Screen name="Main" component={() =>  <Component {...pageProps} />} /> 
          
          </Drawer.Navigator>
        </NavigationContainer>
  )
}
```

**Comment by divonelnc** [comment]
I was running into the same issue and using @jatinderOfc 's trick worked around it. I did some very slight changes to it to remove the `90%` styling and only call setInitRender once:

```
export default function DrawerNavigator() {
  const [initRender, setInitRender] = useState(true);

  useEffect(() => {
    setInitRender(false);
  }, []);

  return (
    <Drawer.Navigator
      drawerContent={(props) => <AppDrawer {...props} />}
      drawerStyle={initRender ? { width: null } : null}
    >
        /*screens*/
    </Drawer.Navigator>
}
```

**Comment by MarkCSmith** [comment]
We tried many of the workarounds suggested in this issue and eventually found that using @divonelnc's solution worked well for us with one change:  we had to replace `width: null` with `height: 0`.

**Comment by vesa193** [comment]
> I was able to fix it by doing this:
> 
> ```js
> export default function DrawerNavigator() {
>   const [initRender, setInitRender] = useState(true);
> 
>   useEffect(() => {
>     setInitRender(false);
>   }, [initRender]);
> 
>   return (
>     <Drawer.Navigator
>       drawerContent={(props) => <AppDrawer {...props} />}
>       drawerStyle={{ width: initRender ? null : "90%" }}
>     >
>         /*screens*/
>     </Drawer.Navigator>
> ```

For this solution works, the difference is

> None of the answers above helped me. I'm working with the React Native Web + Next.js
> Here is how I solved this issue:
> 
> I used the initial render hook @okwast
> 
> ```
> const useInitialRender = (): boolean => {
>   const [isInitialRender, setIsInitialRender] = React.useState(false);
> 
>   if (!isInitialRender) {
>     setTimeout(() => setIsInitialRender(true), 1);
>     return true;
>   }
>   return false;
> };
> ```
> 
> Unfortunately, **width: 0 | null** didn't help so I used this dirty hack. Hope it helps someone.
> 
> ```
> function MyApp({ Component, pageProps }: any) {
>   const isInitialRender = useInitialRender();
>   
>   return (
>         <NavigationContainer>
>           <Drawer.Navigator 
>             openByDefault={false}
>             drawerType={ isInitialRender ? 'permanent' : 'front' } // without it the drawer was blinking even with the _display: 'none'_
>             drawerStyle={{ 
>             width: 240,
>             display: isInitialRender ? 'none' : 'flex' }}
>             drawerContent={(props: any) => <DrawerContent {...props }/>}
>              initialRouteName="Main">
>             <Drawer.Screen name="Main" component={() =>  <Component {...pageProps} />} /> 
>           
>           </Drawer.Navigator>
>         </NavigationContainer>
>   )
> }
> ```

It works like this, i just put `drawerType="back"`, then it behavior nice, on init of my app.

**Comment by pinpointpanda** [comment]
If anyone is using TypeScript and getting a warning (quite rightly!) about setting a `width` property to `null` you can use `undefined` instead of `null` and it seems to work just the same.

**Comment by mariohmol** [comment]
> Closing since no minimal runnable repro was provided! If anyone can provide a repro create a new issue with more info.
> 
> I tested 4 project using react navigation v5 and could not reproduce it. Try upgrading to version 5 maybe your issue will get fixed.

Hi, I did create an repo that reproduces this error!!

Could you please open this or refer to another issue which is trying to fix this issue?

https://github.com/mariohmol/drawertest

Thanks a lot!

**Comment by mariohmol** [comment]
Guys, I could reproduce this error on expo SDK39, after upgrading to SDK40, it is working pretty good!

**Comment by jehartzog** [comment]
I found a hacky way to fix this. works great though. I tried using using `null` for initial render and while we thought that helped, it did not clear the issue. We were still getting flashes of white, but at least it wasn't accidentally grabbing touches. We are still on `react-navigation` v4, so we're stuck w/ static values and unable to dynamically update the width as needed to try the other fix.

Turns out only the width adjustment is necessary, but AFAIK to get this using v4 you need to patch the package to hardcode it in. Here is the (`patch-package`)[https://gist.github.com/jehartzog/1565654ec70bc469a0327ef95d990f17] we used to do it.

Only downside is if you open the drawer before timeout is up, it stays hidden, but it will automatically animate open once timeout is done. Works perfectly for our use case, and you could probably shorten the timeout if needed.

**Comment by qualitymanifest** [comment]
This is still a bug in v6. Setting the `drawerStyle` `width` to 260 causes the drawer to vibrate / wiggle on my coworker's Galaxy S9. It's fine on all iOS devices, fine on Android emulators as well as my Galaxy A12 and my older Alcatel test device.

The fix people discuss here where they conditionally set the `drawerStyle` `width` after the initial render didn't work for me. What ended up working was just removing `drawerStyle` altogether, which was an acceptable solution since I was only setting the width and the default width is good enough.

**Comment by Mitdd9707** [comment]
> I had a similar issue where the drawer flashed upon initial load, and (more problematic) I also saw my navigation drawer actually vibrating back and forth slightly.
> 
> software	version
> react-navigation-drawer	2.2.2
> react-native	0.59.9
> node	8.16.2
> yarn	1.19.0
> As you can see in the gif below, the drawer stutters back and forth. This was noticed only on Android devices, primarily on emulators and devices with 420dpi resolution. The gif example is running on a Pixel 2 Emulator, API v. 27 (Android 8.1 as target).
> 
> My original code for the drawer navigator looked like:
> 
> ```
> const DrawerNavigator = createDrawerNavigator(MainRoutesObj, {
>   drawerWidth: 300
> });
> ```
> 
> ![reactNative_navDrawer_stutters_trimmed](https://user-images.githubusercontent.com/2228826/68514236-2385de80-024b-11ea-8489-6b66d02eb82f.gif)
> 
> Based on conversation above, I removed the drawerWidth, and when I reloaded in the emulator the drawer no longer stuttered back and forth.
> 
> Interestingly enough, I also tried an odd-numbered width for the drawer (301), and that also caused the stuttering to stop.
> 
> Hopefully my additional info helps pinpoint the root cause here. Otherwise, not setting a drawerWidth seems to be a valid workaround.

Thanks @alatzl it works for me. I removed drawer width and its work. now drawer is not blinking
